### ISSUE TITLE
Unable to reproduce PSPNet/FCN8 results

### ISSUE THREAD

**Issue by aicaffeinelife** [yassouali/pytorch_segmentation#29]
Hi, 

Thanks for building this cool library. I've been trying to run some benchmarks and reproduce the results of existing papers from your code. So far I've tested FCN8s and PSPNet and here's what I've obtained:
- FCN8s (VGG backbone): 59.7% mIoU
- PSPNet (Resnet101, dilated convolutions): 71.6% mIoU. 

Here is a snapshot of final evaluation after 100 epochs:
![pspnet_100_val](https://user-images.githubusercontent.com/21975665/69966084-ab29da80-14e3-11ea-9440-d1491e34c925.png)

Here is the config file I've used to train the PSPNet:
```json
{
    "name": "PSPNet",
    "n_gpu": 4,
    "use_synch_bn": true,

    "arch": {
        "type": "PSPNet",
        "args": {
            "backbone": "resnet101",
            "freeze_bn": false,
            "freeze_backbone": false
        }
    },

    "train_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "/path/to/VOC/",
            "batch_size": 8,
            "base_size": 400,
            "crop_size": 380,
            "augment": true,
            "shuffle": true,
            "scale": true,
            "flip": true,
            "rotate": true,
            "blur": false,
            "split": "train_aug",
            "num_workers": 8
        }
    },

    "val_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "/path/to/VOC",
            "batch_size": 8,
            "crop_size": 480,
            "val": true,
            "split": "val",
            "num_workers": 4
        }
    },

    "optimizer": {
        "type": "SGD",
        "differential_lr": true,
        "args":{
            "lr": 0.001,
            "weight_decay": 1e-4,
            "momentum": 0.9
        }
    },

    "loss": "CrossEntropyLoss2d",
    "ignore_index": 255,
    "lr_scheduler": {
        "type": "Poly",
        "args": {}
    },

    "trainer": {
        "epochs": 100,
        "save_dir": "saved/",
        "save_period": 10,
        "monitor": "max Mean_IoU",
        "early_stop": 10,
        "tensorboard": true,
        "log_dir": "saved/runs",
        "log_per_iter": 20,

        "val": true,
        "val_per_epochs": 5
    }
}
```
Following #26, I kept the batch size to 8 and decreased the learning rate to 1e-3. I also checked that the loss correctly weights the auxiliary loss. I'm willing to help debug and reproduce the performance, if you can take a look into it. 

**EDIT**: I am using the augmented Pascal VOC dataset, following the instructions in your Readme and the DeepLabv1 readme. 
Best,

**Comment by yassouali** [comment]
Hi, 

Yes, you're right, the results needs to be higher, the last time I run some experiments, PSPNet with resnet50 (the checkpoint provided), I got 82.5 mIoU (you can see it in the checkpoint). The configs used are as follows :

```json
{'name': 'PSPNet',
 'n_gpu': 2,
 'arch': {'type': 'PSPNet',
  'args': {'backbone': 'resnet50',
   'freeze_bn': False,
   'freeze_backbone': False}},
 'train_loader': {'type': 'VOC',
  'args': {'data_dir': '/workdir/oualiy/Datasets/PascalVoc/VOC',
   'batch_size': 24,
   'base_size': 420,
   'crop_size': 380,
   'augment': True,
   'shuffle': True,
   'scale': True,
   'flip': True,
   'rotate': True,
   'blur': False,
   'split': 'train_aug',
   'num_workers': 8}},
 'val_loader': {'type': 'VOC',
  'args': {'data_dir': '/workdir/oualiy/Datasets/PascalVoc/VOC',
   'crop_size': 480,
   'batch_size': 24,
   'val': True,
   'split': 'val',
   'num_workers': 8}},
 'optimizer': {'type': 'SGD',
  'differential_lr': True,
  'args': {'lr': 0.01, 'weight_decay': 0.0001, 'momentum': 0.9}},
 'loss': 'CrossEntropyLoss2d',
 'ignore_index': 255,
 'lr_scheduler': {'type': 'Poly', 'args': {}},
 'trainer': {'epochs': 100,
  'save_dir': 'saved/',
  'save_period': 5,
  'monitor': 'max Mean_IoU',
  'early_stop': 10,
  'tensorboard': True,
  'log_dir': 'saved/runs',
  'log_per_iter': 20,
  'val': True,
  'val_per_epochs': 5}}
```

So yeah, you can push the batch size a lot more with 4 GPUs (I don't know yours, but the config above was used for two p100 with 24 GB GRAM), you can test with a batch of 32.

**Comment by aicaffeinelife** [comment]
Alright, I can try that and get back to you.

**Comment by aicaffeinelife** [comment]
So, I tried with a batch size of 24 (idea was to go up until atleast 79% mIoU) was obtained but this is where I was at after 100 epochs:

![pspnet_modified](https://user-images.githubusercontent.com/21975665/70003444-21a0f980-1531-11ea-8c2a-e252db0756a6.png)


Reading your config again, I see that you do not use `sync_bn = True` in it, could synchronized batchnorm be the issue?

**Comment by yassouali** [comment]
Actually yes, I added synch bn after those test, can you try and see if that solves the issue?

**Comment by aicaffeinelife** [comment]
Hey, thanks for the quick response. Let me try to turn that to False and get back.

**Comment by aicaffeinelife** [comment]
I tried doing what you suggested. However, it seems that that didn't help

Here is the screenshot of last two evaluations:

![pspnet_no_sbn](https://user-images.githubusercontent.com/21975665/70059178-50a98080-15ae-11ea-8b5e-67033e04cb25.png)

And here is the validation characteristics
![val_characterstics_psp](https://user-images.githubusercontent.com/21975665/70059218-6028c980-15ae-11ea-9ce3-276a1d2a86c8.png)

You can see that the val loss begins diverging after about 4k iterations. I've checked the learning rate and both follow the `poly` decay. I also recall that the authors reported their results with a Resnet101.  What do you think?

**Comment by yassouali** [comment]
The results certainly needs to be higher. Do you start training from scratch ? What is the learning rate you used ? What GPUs ? can you post the last config you tried.

**Comment by aicaffeinelife** [comment]
Training is from scratch with an `lr = 0.01` and `differential_lr =True`. The GPUs are Volta V100 connected over NVLink. This is the config which I tried:

```json
{
    "name": "PSPNet",
    "n_gpu": 4,
    "use_synch_bn": false,

    "arch": {
        "type": "PSPNet",
        "args": {
            "backbone": "resnet50",
            "freeze_bn": false,
            "freeze_backbone": false
        }
    },

    "train_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "/path/to/VOC/",
            "batch_size": 24,
            "base_size": 400,
            "crop_size": 380,
            "augment": true,
            "shuffle": true,
            "scale": true,
            "flip": true,
            "rotate": true,
            "blur": false,
            "split": "train_aug",
            "num_workers": 8
        }
    },

    "val_loader": {
        "type": "VOC",
        "args":{
            "data_dir": "/path/to/VOC",
            "batch_size": 24,
            "crop_size": 480,
            "val": true,
            "split": "val_aug",
            "num_workers": 4
        }
    },

    "optimizer": {
        "type": "SGD",
        "differential_lr": true,
        "args":{
            "lr": 0.01,
            "weight_decay": 1e-4,
            "momentum": 0.9
        }
    },

    "loss": "CrossEntropyLoss2d",
    "ignore_index": 255,
    "lr_scheduler": {
        "type": "Poly",
        "args": {}
    },

    "trainer": {
        "epochs": 120,
        "save_dir": "saved/",
        "save_period": 10,
        "monitor": "max Mean_IoU",
        "early_stop": 10,
        "tensorboard": true,
        "log_dir": "saved/runs",
        "log_per_iter": 20,

        "val": true,
        "val_per_epochs": 5
    }
}

```

**Comment by yassouali** [comment]
I'll launch some experiments and I'll get back to you

**Comment by aicaffeinelife** [comment]
Any update @yassouali?

**Comment by yassouali** [comment]
I run some tests using your config, and I got the same results, I'll try to find the problem ASAP

**Comment by aicaffeinelife** [comment]
I'm running some experiments with some of my own intuitions and ideas. I'll get back to you with them and if I can reproduce the results, I'll share the config.

**Comment by huqiming513** [comment]
Mine is worse than that with the following config trained from scratch.
The GPU is RTX 2080Ti and my available device is only one.  
```
{
  "name": "PSPNet",
  "n_gpu": 1,
  "arch": {
    "type": "PSPNet",
    "args": {
      "backbone": "resnet50",
      "freeze_bn": false,
      "freeze_backbone": false
    }
  },
  "train_loader": {
    "type": "VOC",
    "args": {
      "data_dir": "/home/huqiming/datasets",
      "batch_size": 8,
      "base_size": 420,
      "crop_size": 380,
      "augment": true,
      "shuffle": true,
      "scale": true,
      "flip": true,
      "rotate": true,
      "blur": false,
      "split": "trainval_aug",
      "num_workers": 8
    }
  },
  "val_loader": {
    "type": "VOC",
    "args": {
      "data_dir": "/home/huqiming/datasets",
      "crop_size": 480,
      "batch_size": 8,
      "val": true,
      "split": "val",
      "num_workers": 8
    }
  },
  "optimizer": {
    "type": "SGD",
    "differential_lr": true,
    "args": {
      "lr": 0.01,
      "weight_decay": 1e-4,
      "momentum": 0.9
    }
  },
  "loss": "CrossEntropyLoss2d",
  "ignore_index": 255,
  "lr_scheduler": {
    "type": "Poly",
    "args": {}
  },
  "trainer": {
    "val_on_start": false,
    "epochs": 200,
    "save_dir": "saved/",
    "save_period": 10,
    "monitor": "max Mean_IoU",
    "if_early_stop": false,
    "early_stop": 10,
    "tensorboard": true,
    "log_dir": "saved/runs",
    "log_per_iter": 20,
    "show_per_batch": 5,
    "val": true,
    "val_per_epochs": 1
  }
}
```
![image](https://user-images.githubusercontent.com/31566437/70384607-2dbb0a00-19bc-11ea-84ee-34de32fd1fb7.png)

**Comment by aicaffeinelife** [comment]
Hey @huqiming513, try with a bigger batch size

**Comment by aicaffeinelife** [comment]
Hey @huqiming513, try with a bigger batch size

**Comment by huqiming513** [comment]
Hi @aicaffeinelife, I used the pretrained backbone and the results were hugely improved, the config of which is the same as above but the training batch size is only 4. If you also used the pretrained backbone, the resnet50 backbone may be a better choice. :)  
![image](https://user-images.githubusercontent.com/31566437/70842768-9265e100-1e62-11ea-9762-a4a27c505798.png)

**Comment by aicaffeinelife** [comment]
Which pre-trained backbone? The one provided by official PyTorch one?

**Comment by aicaffeinelife** [comment]
Okay interesting results @yassouli - training with *trainval* on Pascal VOC leads to substantially higher results. I've only tried with single GPU, but I'll keep the thread open for other findings.

**Comment by aicaffeinelife** [comment]
Okay some interesting results, I got substantially high results when I trained with _trainval_ set and evaluated on Pascal VOC. Tested on only one GPU with a batch size of 4. I'm gonna keep the thread open with some other results.

**Comment by Dongshengjiang** [comment]
I got similar accuracy of 0.71 Mean_IoU for the default config provided by gitgub code. Could you please attach the config you used to get the highest accurcy? Another issue is how about the results of using UNet and other networks. It would be nice to show their results in your readme.

**Comment by yassouali** [comment]
Unfortunately right now I am quite busy. Hopefully soon I will retrain and post the results for the other model / datasets and the optimal config for each model. Until then, let's keep this issue open.

**Comment by Dongshengjiang** [comment]
thanks dongshengjiang 邮箱：dongshengjiang@aliyun.com 签名由 网易邮箱大师 定制 On 01/20/2020 23:53, Yassine wrote: Unfortunately right now I am quite busy. Hopefully soon I will retrain and post the results for the other model / datasets and the optimal config for each model. Until then, let's keep this issue open. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by iwyoo** [comment]
@yassouali 
Hi, Is there any update on this issue? If PSPNet performance was reproduced well before, it seems that we can compare the old code commits. Is the performance table in the README reproduced by using this repository?

**Comment by yassouali** [comment]
closed for now, can be opened later
